### PR TITLE
adding pnp

### PR DESCRIPTION
**Pull request by edgarriba**
the code is running and solving. Need to test the performance with good training

### CODE REVIEW

**Comment by vrabaud** [comment]
please rebase, this code cannot be merged as is

**Comment by edgarriba** [comment]
I think it's okay right now

File: python/object_recognition_tod/trainer.py [hunk 1]
```diff
@@ -4,8 +4,8 @@
 """
 
 from ecto import BlackBoxCellInfo as CellInfo, BlackBoxForward as Forward
-from ecto_opencv import calib, features2d, highgui
-from ecto_opencv.features2d import FeatureDescriptor
+#from ecto_opencv import calib, features2d, highgui
```
Comment (vrabaud) [inline python/object_recognition_tod/trainer.py]: remove if not needed

File: python/object_recognition_tod/trainer.py [hunk 2]
```diff
@@ -42,7 +43,17 @@ def declare_forwards(cls, _p):
...
         return (p,i,o)
 
-    def connections(self, p):
+    @classmethod
+    def declare_direct_params(self, p):
+        p.declare('feature', 'The features parameters as a JSON string', '{}')
+        p.declare('descriptor', 'The descriptor parameters as a JSON string', '{}')
+
+    def configure(self, p, _i, _o):
+      self.trainer = ecto_training.Trainer("Trainer",
```
Comment (vrabaud) [inline python/object_recognition_tod/trainer.py]: respect the indentation

File: src/detection/DescriptorMatcher.cpp [hunk 3]
```diff
@@ -196,19 +205,15 @@ namespace tod
...
       {
-        if (matcher_->getTrainDescriptors().empty())
-        {
-          std::cerr << "No descriptors loaded" << std::endl;
-          return ecto::OK;
-        }
         // Perform radius search
         // As this does not work for LSH on OpenCV 2.4, we first perform knn
-        matcher_->knnMatch(descriptors, matches, 5);
+        /*matcher_->knnMatch(descriptors, matches, 5);
```
Comment (vrabaud) [inline src/detection/DescriptorMatcher.cpp]: remove if not necessary

File: src/detection/DescriptorMatcher.cpp [hunk 4]
```diff
@@ -217,14 +222,30 @@ namespace tod
...
+      else
+      {
+        std::cerr << "No descriptors loaded" << std::endl;
+        return ecto::OK;
       }
 
-      // TODO Perform ratio testing if necessary
-      if (ratio_)
+      // Perform 2 Nearest Neighbors search
+      matcher_->knnMatch(descriptors, matches, 2);
```
Comment (vrabaud) [inline src/detection/DescriptorMatcher.cpp]: why only 2 ?

File: src/detection/GuessGenerator.cpp [hunk 5]
```diff
@@ -139,22 +141,136 @@ namespace tod
...
       {
-        // Only use 2d to 3d matching
-        // TODO
-        //const std::vector<cv::KeyPoint> &keypoints = inputs.get<std::vector<cv::KeyPoint> >("keypoints");
+        std::cout << "*** PnP METHOD ***" << std::endl;
+
+        // -- pnp config
+
+        // distortion coefficients
+        cv::Mat dist_coef = cv::Mat::zeros(1, 4, CV_32F);
```
Comment (vrabaud) [inline src/detection/GuessGenerator.cpp]: use a cv::Matx14f

File: src/training/ModelFiller.cpp [hunk 6]
```diff
@@ -1,5 +1,12 @@
 #include <ecto/ecto.hpp>
-#include <opencv2/core/core.hpp>
+
+
+#if CV_VERSION_MAJOR < 3
```
Comment (vrabaud) [inline src/training/ModelFiller.cpp]: the proper way is to look at the existence of CV_VERSION_EPOCH or not (it's not in OpenCV3) and yes, that needs to be changed in all OpenCV ROS code (for now, I create a pre-processor variable OPENCV3)

File: src/training/Trainer.cpp [hunk 7]
```diff
@@ -33,30 +33,28 @@
...
 
-void rescale_depth(const cv::Mat depth_in, const cv::Size & isize,
-    cv::Mat &depth_out) {
+void
+rescale_depth(const cv::Mat depth_in, const cv::Size & isize, cv::Mat &depth_out)
+{
   cv::Size dsize = depth_in.size();
   cv::Mat depth;
-  rescaleDepth(depth_in, CV_32F, depth);
+  cv::rescaleDepth(depth_in, CV_32F, depth);
```
Comment (vrabaud) [inline src/training/Trainer.cpp]: that's in <opencv2/rgbd/rgbd.hpp> , you need the header

File: src/training/Trainer.cpp [hunk 8]
```diff
@@ -93,48 +97,108 @@ struct Trainer {
...
+  static void 
+  declare_io(const ecto::tendrils& params, ecto::tendrils& inputs, ecto::tendrils& outputs)
+  {
+    // Inputs
     inputs.declare(&Trainer::json_db_, "json_db",
         "The parameters of the DB as a JSON string.").required(true);
     inputs.declare(&Trainer::object_id_, "object_id",
         "The id of the object in the DB.").required(true);
 
+    // Outpus
```
Comment (vrabaud) [inline src/training/Trainer.cpp]: outputs

**Comment by vrabaud** [comment]
please fix problems, squash a few commits so that we understand better what is happening, and make sure it compiles with OpenCV2

File: src/detection/DescriptorMatcher.cpp (hunk 4 above)
Comment (edgarriba) [inline src/detection/DescriptorMatcher.cpp]: for me was working in this way. You think it's better with more?

File: src/training/ModelFiller.cpp (hunk 6 above)
Comment (edgarriba) [inline src/training/ModelFiller.cpp]: are you doing that?

Comment (vrabaud) [inline src/training/ModelFiller.cpp]: I've started in opencv_candidate.

Comment (edgarriba) [inline src/training/ModelFiller.cpp]: ok. I've seen in ork core there is already this pre-processor flag

Comment (vrabaud) [inline src/training/ModelFiller.cpp]: this is now the only proper way of checking for OpenCV2 / 3: https://github.com/wg-perception/opencv_candidate/blob/master/src/opencv_candidate/feature2d.cpp#L113

Comment (edgarriba) [inline src/training/ModelFiller.cpp]: okay, but I'm using 2.4.11 and CV_VERSION_EPOCH is missing.

**Comment by vrabaud** [comment]
Are you sure ? it's in trunk of 2.4:
https://github.com/Itseez/opencv/blob/2.4/modules/core/include/opencv2/core/version.hpp#L50
Major does not have the same meaning in 2.4 and 3:
https://github.com/Itseez/opencv/blob/2.4/modules/core/include/opencv2/core/version.hpp#L50
https://github.com/Itseez/opencv/blob/master/modules/core/include/opencv2/core/version.hpp#L53

**Comment by edgarriba** [comment]
yeah, it's weird. As you prefer, my solution works as well.

**Comment by vrabaud** [comment]
Here is the situation:
https://github.com/Itseez/opencv/blob/master/modules/core/include/opencv2/core/version.hpp
https://github.com/Itseez/opencv/blob/2.4/modules/core/include/opencv2/core/version.hpp

CV_VERSION_EPOCH does not exist in 3.0, hence you can only check for its existence (that's what I advise you to do). CV_VERSION_MAJOR has a totally different meaning between 2.4 and 3.x and should therefore not be used (there might be an OpenCV 2.5, who knows).

Then again, all that might change before the 3.x release :)

File: CMakeLists.txt [hunk 9]
```diff
@@ -20,6 +20,13 @@ find_package(Boost REQUIRED system filesystem serialization)
 find_package(Eigen REQUIRED)
 find_package(OpenCV REQUIRED)
 
+if ( OpenCV_VERSION VERSION_GREATER 2.4.9 )
+ add_definitions("-DOPENCV3=1")
```
Comment (vrabaud) [inline CMakeLists.txt]: that's what I do in other packages yes but I told you the right thing to do is simply to check for the existence of the EPOCH variable in the header and that's it: this is lighter that adding a pre-processor variable.

**Comment by edgarriba** [comment]
I understand but then I cannot compile my code since I don't know what's wrong in my computer with this flag. I tried to uninstall and install opencv twice but it's the same. Any idea?

**Comment by vrabaud** [comment]
since you don't give any error message, no :)
DO you have an OpenCV in /usr/lib ? one in /usr/local/lib ? Any other one ? Display OpenCV_INCLUDE_DIRS in your CMake and you'll know which OpenCV it's using.

**Comment by edgarriba** [comment]
I do only have opencv in /usr/local/lib.

My cmake:

-- OpenCV version: 2.4.11
-- OpenCV headers: /usr/local/include/opencv;/usr/local/include


Then if I put #ifndef CV_VERSION_EPOCH in my code I get:

/home/eriba/catkin_ws/src/ork/tod/src/detection/DescriptorMatcher.cpp:46:30: fatal error: opencv2/core.hpp: El fitxer o directori no existeix
   #include <opencv2/core.hpp>
                              ^
compilation terminated.
make[2]: *** [ork/tod/src/CMakeFiles/ecto_detection_ectomodule.dir/detection/DescriptorMatcher.cpp.o] Error 1
make[1]: *** [ork/tod/src/CMakeFiles/ecto_detection_ectomodule.dir/all] Error 2
make: *** [all] Error 2
Invoking "make -j 1" failed

**Comment by vrabaud** [comment]
well, CV_VERSION_EPOCH is only defined after including opencv2/core/core.hpp
That's why you should first do a ```#include <opencv2/core/version.hpp>``` and then use that ifndef

**Comment by vrabaud** [comment]
Please rebase. You also have instructions for OpenCV3 here: http://wiki.ros.org/opencv3

**Comment by vrabaud** [comment]
any way you can squash some commits too please ? Thx !

**Comment by edgarriba** [comment]
sure, but it will take a while since this was some time ago and now I need to reinstall all the ROS stuff again

**Comment by vrabaud** [comment]
well, rebasing and squashing do not require to compile anything :)

**Comment by edgarriba** [comment]
oki but then you'll need to check that it really works

**Comment by vrabaud** [comment]
sure, I just want to simplify your pull request by cleaning the OpenCV3 stuff. TOD finally compiles with it.

**Comment by edgarriba** [comment]
now it's rebased. About squashing commits, I enter in a never ending loop of merge conflicts. Any idea how to solve easily?

**Comment by vrabaud** [comment]
well, only squash consecutive commits for now: git rebase -i and put an 's' for the squashable ones. All the merges, the commits with the same message, with the same purpose (OpenCV3 compatibility for example). A utility like qgit to visualize quickly the content of each commit could help you out. Anyway, thx, I can start looking at it now. Thx !

File: src/common/adjacency_ransac.cpp [hunk 10]
```diff
@@ -36,14 +36,16 @@
...
 
-#include <opencv2/core/core.hpp>
-#include <opencv2/features2d/features2d.hpp>
-#include <opencv2/highgui/highgui.hpp>
-
 #include "adjacency_ransac.h"
 #include "ransac.h"
 #include "sac_model_registration_graph.h"
 
+#ifndef CV_VERSION_EPOCH
```
Comment (vrabaud) [inline src/common/adjacency_ransac.cpp]: The latest OpenCV3 should be fine with that file, you don't need to modify it. Thx

File: src/common/adjacency_ransac.h [hunk 11]
```diff
@@ -39,10 +39,12 @@
 #include <map>
 #include <vector>
 
-#include <opencv2/core/core.hpp>
-
```
Comment (vrabaud) [inline src/common/adjacency_ransac.h]: The latest OpenCV3 should be fine with that file, you don't need to modify it. Thx

File: src/detection/GuessGenerator.cpp [hunk 12]
```diff
@@ -42,10 +42,19 @@
 
 #include <ecto/ecto.hpp>
 
-#include <opencv2/core/core.hpp>
-#include <opencv2/imgproc/imgproc.hpp>
-#include <opencv2/features2d/features2d.hpp>
-#include <opencv2/highgui/highgui.hpp>
+#include <opencv2/core/version.hpp>
```
Comment (vrabaud) [inline src/detection/GuessGenerator.cpp]: The old headers should work.

File: src/training/ModelFiller.cpp [hunk 13]
```diff
@@ -1,4 +1,5 @@
 #include <ecto/ecto.hpp>
+
```
Comment (vrabaud) [inline src/training/ModelFiller.cpp]: Don't modify that file

File: src/training/training.cpp [hunk 14]
```diff
@@ -33,22 +33,22 @@
...
-#include <opencv2/rgbd.hpp>
-namespace cv {
-using namespace cv::rgbd;
-}
-#else
-#include <opencv2/rgbd/rgbd.hpp>
-#endif
-
-#include "training.h"
+ #include <limits.h>
```
Comment (vrabaud) [inline src/training/training.cpp]: I don't think that file needs to be modified.

File: src/training/training.h [hunk 15]
```diff
@@ -33,12 +33,24 @@
...
+#include <opencv2/core/version.hpp>
+
+#ifndef CV_VERSION_EPOCH
+  #include <opencv2/imgproc.hpp>
+#else
+  #include <opencv2/imgproc/imgproc.hpp>
+  #include <opencv2/features2d/features2d.hpp>
+#endif
+
+#include <opencv2/rgbd/rgbd.hpp>
```
Comment (vrabaud) [inline src/training/training.h]: That would not work with OpejnCV 3.1. The previous headers would though.

**Comment by edgarriba** [comment]
@vrabaud in 04e5aef seems to be all the commits, but not sure if I did it well.

**Comment by vrabaud** [comment]
hmmm, at this point, it might be easier to copy the files over a French and make one commit. Your history is non-linear which will make merging pretty tough .... Sorry/

**Comment by edgarriba** [comment]
Copy where? :-D
close this once you check https://github.com/wg-perception/tod/pull/13

**Comment by edgarriba** [comment]
@vrabaud  should we close this?

**Pull request closed without merging**